passes/opt/opt_expr.cc

```diff
@@ -512,10 +513,14 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
 			cells.edge(cells.node(outbit_to_cell.at(bit)), r_index);
 	}
 
-	if (!cells.sort()) {
+	if (sort_fails < effort && !cells.sort()) {
 		// There might be a combinational loop, or there might be constants on the output of cells. 'check' may find out more.
 		// ...unless this is a coarse-grained cell loop, but not a bit loop, in which case it won't, and all is good.
 		log("Couldn't topologically sort cells, optimizing module %s may take a longer time.\n", log_id(module));
+		sort_fails++;
+		if (sort_fails >= effort)
+			log("Effort of %d exceeded, no longer attempting toposort on module %s.\n",
+				 effort, log_id(module));
 	}
 
 	for (auto cell : cells.sorted)
```

Comment:
Isn't this empty if you don't run `cells.sort()`? And if that's the case, then the true cause of the speed-up isn't not bothering to sort anymore but giving up entirely

Reply:
Fixed by iterating over a list of references to a snapshot of `module->cells()`. This increases the runtime to 83 seconds, which I still prefer to 326 seconds. I should say, all runtimes I'm mentioning in this PR are measured with `ENABLE_DEBUG=1` for no particular reason other having that around in `Makefile.conf`

Reply:
We could cache the sorting result provided we remove cells we removed

Reply:
I can't really foresee the consequences of doing that, though - for how many cycles would we cache etc

Reply:
For the lifetime of one `opt_expr` invocation. It should be ok as long as we zero out the position in the sorted array corresponding to any removed cells.

I am not sure if there's a case where opt_expr replaces a cell with two or more cells, that's a slight complication